SpatialGDK/Source/SpatialGDK/Public/SpatialView/OutgoingComponentMessage.h

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+#include "SpatialView/ComponentData.h"
+#include "SpatialView/ComponentUpdate.h"
+#include "SpatialView/EntityComponentId.h"
+
+namespace SpatialGDK
+{
+
+// Represents one of a component addition, update, or removal.
+// Interally schema data is stored using raw pointers. However the interface exclusively uses explicitly owning objects to denote ownership.
+// Not safe to use after being moved from.
+class OutgoingComponentMessage
+{
+public:
+	enum MessageType {ADD, UPDATE, REMOVE};
+
+	explicit OutgoingComponentMessage(Worker_EntityId EntityId, ComponentData ComponentAdded)
+		: EntityId(EntityId), ComponentId(ComponentAdded.GetComponentId()), ComponentAdded(MoveTemp(ComponentAdded).Release()), Type(ADD)
+	{
+	}
+
+	OutgoingComponentMessage(Worker_EntityId EntityId, ComponentUpdate ComponentUpdated)
+		: EntityId(EntityId), ComponentId(ComponentUpdated.GetComponentId()), ComponentUpdated(MoveTemp(ComponentUpdated).Release()), Type(UPDATE)
+	{
+	}
+
+	OutgoingComponentMessage(Worker_EntityId EntityId, Worker_ComponentId RemovedComponentId)
+		: EntityId(EntityId), ComponentId(RemovedComponentId), Type(REMOVE)
+	{
+	}
+
+	~OutgoingComponentMessage()
+	{
+		// As data is stored in owning raw pointers we need to make sure resources are realsed.
+		switch (Type)
+		{
+		case ADD:
+			if (ComponentAdded)
+			{
+				Schema_DestroyComponentData(ComponentAdded);
+			}
+			break;
+		case UPDATE:
+			if (ComponentUpdated)
+			{
+				Schema_DestroyComponentUpdate(ComponentUpdated);
+			}
+			break;
+		case REMOVE:
+			break;
+		}
+	}
+
+	// Moveable, not copyable.
+	OutgoingComponentMessage(const OutgoingComponentMessage&) = delete;
+	OutgoingComponentMessage(OutgoingComponentMessage&& Other) = default;
```

Comment:
I expect the default behaviour to leave the object in its original state, leading to double-deletion.

Reply:
Ah good spot there was meant to be a NONE type in here but we forgot to implement it :P.
In the worker wrapper repo I actually did the other way where there is a union with unique_ptr in it.
I think there was a break when implementing this and I was still going back and forth on whether or not this actually makes it simpler when we add the NONE type and implement move constructor/assignement.

Reply:
Will add the NONE type in to fix this now.